services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/JwtSigningService.java

```diff
@@ -29,14 +42,71 @@
+    private static final Logger LOGGER = Logger.getLogger(JwtSigningService.class);
+
+    private static final String ID_TEMPLATE = "urn:uuid:%s";
+    private static final String TOKEN_TYPE = "JWT";
+    private static final String VC_CLAIM_KEY = "vc";
+    private static final String ID_CLAIM_KEY = "id";
+
+
+    private final SignatureSignerContext signatureSignerContext;
+    private final TimeProvider timeProvider;
+    protected final String issuerDid;
+
+    public JwtSigningService(KeycloakSession keycloakSession, String keyId, String algorithmType, String issuerDid, TimeProvider timeProvider) {
         super(keycloakSession, keyId, algorithmType);
+        this.issuerDid = issuerDid;
+        this.timeProvider = timeProvider;
+        KeyWrapper signingKey = getKey(keyId, algorithmType);
+        if (signingKey == null) {
+            throw new SigningServiceException(String.format("No key for id %s and algorithm %s available.", keyId, algorithmType));
+        }
+        signatureSignerContext = switch (algorithmType) {
+            case Algorithm.RS256, Algorithm.RS384, Algorithm.RS512, Algorithm.PS256, Algorithm.PS384, Algorithm.PS512 ->
+                    new AsymmetricSignatureSignerContext(signingKey);
+            case Algorithm.ES256, Algorithm.ES384, Algorithm.ES512 -> new ECDSASignatureSignerContext(signingKey);
+            default ->
+                    throw new SigningServiceException(String.format("Algorithm %s is not supported by the JWTSigningService.", algorithmType));
+        };
+        LOGGER.debugf("Successfully initiated the JWT Signing Service with algorithm %s.", algorithmType);
     }
 
     @Override
     public String signCredential(VerifiableCredential verifiableCredential) {
+        LOGGER.debugf("Sign credentials to jwt-vc format.");
+        // set mandatory fields
+        JsonWebToken jsonWebToken = new JsonWebToken()
+                .issuer(verifiableCredential.getIssuer().toString())
+                .nbf((long) timeProvider.currentTime())
+                .iat((long) timeProvider.currentTime())
+                .id(createCredentialId(verifiableCredential));
+        jsonWebToken.setOtherClaims(VC_CLAIM_KEY, verifiableCredential);
+
+        // expiry is optional
+        Optional.ofNullable(verifiableCredential.getExpirationDate())
+                .ifPresent(d -> jsonWebToken.exp(d.toInstant().getEpochSecond()));
 
-        throw new UnsupportedOperationException("JWT-VC Credentials Signing is not yet supported.");
+        // subject id should only be set if the credential subject has an id.
+        Optional.ofNullable(
+                        verifiableCredential
+                                .getCredentialSubject()
+                                .getClaims()
+                                .get(ID_CLAIM_KEY))
+                .map(Object::toString)
+                .ifPresent(jsonWebToken::subject);
+
+        return new JWSBuilder()
+                .type(TOKEN_TYPE)
```

Comment:
The token type shall be read from SigningServiceProvider.

Reply:
Can the "typ" claim in the JWT header be something else than "JWT" for a jwt_vc? I thought its mandatory, but I'm not able to find it in the spec anymore.

Reply:
[sd-jwt](https://datatracker.ietf.org/doc/draft-ietf-oauth-selective-disclosure-jwt/06/) is playing around with ```typ: "kb+jwt"```. As the whole is still a moving target, having typ externalized is IMHO a good idea.

Reply:
I know sd-jwt defines another header. What do you think about introducing this as a property of the ServiceProvider(same as with KeyId and Algorithm Type) and make JWT just the default. Would this be enough flexibility?

Reply:
See [5270563](https://github.com/keycloak/keycloak/pull/26484/commits/5270563f632f0f9406e0c275aa9f16132409bd54)